### PR TITLE
Add FSType for CSI volume source

### LINKED ISSUE #58183: CSI - fsType is not being passed to driver
/kind bug

**What happened**:
Unable to specify fsType from a CSI PV.  The fsType is currently hardcoded to ext4.

**What you expected to happen**:
Being able to specify fsType when creating a CSI PV
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: manually-created-pv
spec:
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteOnce
  persistentVolumeReclaimPolicy: Retain
  csi:
    driver: com.example.team/csi-driver
    volumeHandle: existingVolumeName
    readOnly: false
    fsType: ext4
```

### PR DESCRIPTION
**Pull request by NickrenREN**
**What this PR does / why we need it**:
Add FSType for CSI volume source to specify filesystems (alpha defaults to `ext4`)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58183

**Special notes for your reviewer**:

**Release note**:
```release-note
Add FSType for CSI volume source to specify filesystems
```

/assign @saad-ali 
cc @vladimirvivien

### CODE REVIEW

**Comment by liggitt** [comment]
cc @kubernetes/sig-storage-api-reviews

**Comment by jsafrane** [comment]
/approve
from storage point of view

File: pkg/volume/csi/csi_mounter.go [hunk 1]
```diff
@@ -197,7 +197,7 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 		accessMode,
 		c.volumeInfo,
 		attribs,
-		"ext4", //TODO needs to be sourced from PV or somewhere else
+		csiSource.FSType,
```
Comment (jsafrane) [inline pkg/volume/csi/csi_mounter.go]: Comment above says that `Implicitly inferred to be "ext4" if unspecified.` We need some code for that if `csiSource.FSType` is empty.

Comment (NickrenREN) [reply pkg/volume/csi/csi_mounter.go]: I saw that in `drivers` repo, we would defaults `FSType` to `ext4`  in function `formatAndMount` if it was not set.  Should we deal with this here ?
btw, when this gets merged, i will update the external provisioner to add `FSType` for newly created PV if it is set in StorageClass parameters.

Comment (jsafrane) [reply pkg/volume/csi/csi_mounter.go]: We don't have control about all CSI drivers in the wold to fall back to ext4 when nothing is specified.  Other drivers may choose a different default, so it's Kubernetes who should enforce ext4 here.

Comment (NickrenREN) [reply pkg/volume/csi/csi_mounter.go]: done, thanks, PTAL

**Comment by vladimirvivien** [comment]
@NickrenREN last time I had to manually make type changes in staging/src/k8s.io/api/core/v1/types.go . You will need to make FSType change there as well.

**Comment by NickrenREN** [comment]
@vladimirvivien  i have already changed that in my first commit

**Comment by saad-ali** [comment]
/lgtm
/approve

One more thing to consider for API review:

Does this aligns with how we plan to expose block storage long term? Having the `fstype` inside the volume source, if I remember correctly was an API choice that we somewhat regretted when trying to introduce block volume support. That said, all the in-tree volumes currently have fstype as part of the volume source, so this aligns with that.

@mtanino, @erinboyd any comments??

**Comment by msau42** [comment]
During the block design, the idea came up that it could be cleaner if fstype was in the PVC instead of per volume type, so that we could standardize fs handling across all volume types instead of having each plugin handle it.

**Comment by mtanino** [comment]
@saad-ali @msau42 

IIUC,

>> During the block design, the idea came up that it could be cleaner if fstype was in the PVC instead of per volume type, so that we could standardize fs handling across all volume types instead of having each plugin handle it.

Do you mean `volume source` rather than `volume type`?
I believe each volume source or persistent volume source should have fstype since this parameter is characteristic of volume.
But adding fstype into PVC then just pass through the value to PV would be interesting.
because;
- volumeMode and fstype are strongly related parameter
- we can easily validate volumeMode and fstype if both of them are in the same PVC object

However, I believe changing this affects big(?) impact to all plugins, also we need keep backward compatibility, therefore I think we don't have a plan to change this.

@erinboyd @screeley44 
Any comments?

**Comment by msau42** [comment]
Yeah I meant putting the fsType in the higher level PV/PVC, instead of as a parameter in each volume-specific source.  Then fs formatting and supporting scenarios like fs-on-block could be done by Kubernetes instead of every single plugin.

But yes, this is a major change and would be messy when also considering backwards compatibility, so we just stayed with the current way of specifying fsType in every volume source spec.

**Comment by mtanino** [comment]
@saad-ali 
Thus I agree this API change.

**Comment by NickrenREN** [comment]
kindly ping @saad-ali  @msau42  @thockin  for approval

**Comment by saad-ali** [comment]
Sounds good to me. Will need to clean this up for Kubernetes 2.0!

@thockin for API approval

**Pull request merged**